addons/test_mail/tests/test_attachment.py

```diff
@@ -0,0 +1,72 @@
+        admin_attachments = self.env['ir.attachment'].with_user(user_admin).create([{
+            'type': 'binary',
+            'name': f'admin_attachment_{n}',
+            'datas': "ABC=",
+        } for n in range(2)])
+        employee_attachments = self.env['ir.attachment'].with_user(user_employee).create([{
+            'type': 'binary',
+            'name': f'employee_attachment_{n}',
+            'datas': "ABC=",
+        } for n in range(3)])
+
+        user_employee_clone = user_employee.copy(default={'login': 'employee_clone', 'email': 'employee_clone@local.lan'})
+        employee_clone_attachments = self.env['ir.attachment'].with_user(user_employee_clone).create({
+            'type': 'binary',
+            'name': 'employee_clone_attachment',
+            'datas': "ABC=",
+        })
+
+        test_record.with_user(user_admin).message_post(body="Hi", attachment_ids=admin_attachments.ids)
+        test_record.with_user(user_employee).message_post(body="Hello", attachment_ids=employee_attachments.ids)
+        test_record.with_user(user_employee_clone).sudo().message_post(body="Hello again", attachment_ids=employee_attachments[2].ids)
+        create_by_employee_send_by_clone = employee_attachments[2]
+        test_record.with_user(user_employee_clone).sudo().message_post(body="Hello again with own attachment", attachment_ids=employee_clone_attachments.ids)
+
+        # forbidden
+        forbidden_list = [
+            (user_employee, admin_attachments[0], False),
+            (user_employee, employee_clone_attachments[0], False)
+        ]
+        for method in (write, unlink):
+            for user, attachment, sudo in forbidden_list:
+                with self.subTest(user=user.name, attachment=attachment.name, sudo=sudo, method=method.__name__):
+                    with self.assertRaises(AccessError):
+                        method(attachment, user, sudo)
+
+        # allowed
+        allowed_list = [
+            (user_admin, admin_attachments[0], False),
+            (user_admin, employee_attachments[0], False),
+            (user_employee, admin_attachments[1], True),
+            (user_employee, employee_attachments[1], False),
+            (user_employee, create_by_employee_send_by_clone, False),
+        ]
+        for user, attachment, sudo in allowed_list:
+            with self.subTest(user=user.name, attachment=attachment.name, sudo=sudo):
+                write(attachment, user, sudo)
+        self.assertListEqual(admin_attachments.mapped('name'), ['Mitchell Admin 0', 'Ernest Employee 1'])
+        self.assertListEqual(employee_attachments.mapped('name'), ['Mitchell Admin 0', 'Ernest Employee 1', 'Ernest Employee 2'])
+        for user, attachment, sudo in allowed_list:
+            with self.subTest(user=user.name, attachment=attachment.name, sudo=sudo):
+                unlink(attachment, user, sudo)
+        self.assertListEqual(admin_attachments.exists().mapped('name'), [])
```

Comment:
Why not simply checking we have a void recordset, like
```suggestion
        self.assertFalse(admin_attachments.exists())
```

Reply:
Idea for this is that if it's not deleted for some reason you directly know which one is left
But not necessary indeed